### PR TITLE
Add type annotations

### PR DESCRIPTION
**Pull request by bluetech**
This adds type annotations to all pluggy code. This is a respin of #225 but now using Python 3 syntax.

The type annotations are not published, it will require some work for it to make sense, cleaning up the interface and imports and such. But I think internal type annotations are still useful for understand the code (I always have to relearn how stuff relate to each other when looking at pluggy) and as a linter.

In the meantime @RonnyPfannschmidt created #326. I think this PR is a better starting point because it almost doesn't change the code itself besides adding the types. Ronny's changes can be added on top maybe.

### CODE REVIEW

**Review by RonnyPfannschmidt — APPROVED** [review]
Love the approach,

I added types with the goal of mypyc speedup experiments

Those failed

**Review by nicoddemus — APPROVED** [review]
Great work, thanks @bluetech!

**Comment by nicoddemus** [comment]
Took the liberty of committing an update to coverage configuration to ignore typing-related code (c5d57862985bd25752b1c6eae88f56fd1fa4455f).

File: pyproject.toml [hunk 1]
```diff
@@ -45,3 +45,24 @@ template = "changelog/_template.rst"
   directory = "trivial"
   name = "Trivial/Internal Changes"
   showcontent = true
+
+[tool.mypy]
+mypy_path = "src"
+check_untyped_defs = true
+# disallow_any_expr = true
```
[the remark this answers is not in the document]
Comment (bluetech) [reply pyproject.toml:53]: Added a comment

**Comment by bluetech** [comment]
Thanks for the reviews! I think this is good to go.

**Pull request merged**